### ISSUE TITLE
strict-type-predicates reports 'Expression is always true.' which is not true

### ISSUE THREAD

**Issue by Martin-Wegner** [palantir/tslint#2783]
### Bug Report

- __TSLint version__: 5.2.0
- __TypeScript version__: 2.3.2
- __Running TSLint via__: ng lint --type-check from Angular CLI

#### TypeScript code being linted

```ts
['a', 'b', 'c'].map((s: string) => {
  if (s === 'b') {
    return undefined;
  }
  return 1;
}).filter((n: number) => n !== undefined);
```

```ts
['a', 'b', 'c'].map((s: string) => {
  if (s === 'b') {
    return undefined;
  }
  return 1;
}).filter((n: number | undefined) => n !== undefined);
```

```ts
['a', 'b', 'c'].map((s: string) => {
  if (s === 'b') {
    return undefined;
  }
  return 1;
}).filter(n => n !== undefined);
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "strict-type-predicates": true
  }
}

```

#### Actual behavior
> ERROR: ...: Expression is always true.

#### Expected behavior
No error.

**Comment by ajafff** [comment]
Remove the type annotation `:number` which is wrong anyway. That should
make it work

Am 18.05.2017 17:19 schrieb "Martin-Wegner" <notifications@github.com>:

**Comment by Martin-Wegner** [comment]
Why is `number` wrong? The mapping function maps `string` to `number`... When I use `any` the same error occurs.

**Comment by ajafff** [comment]
Your function maps to `number | undefined `. Either use that as type annotation of the parameter or just omit it and let typescript infer the type. 

The bug with `any` is fixed by a recent PR

**Comment by Martin-Wegner** [comment]
I omitted it and the same error occurs...

**Comment by Martin-Wegner** [comment]
Using `number | undefined` also not work :(

**Comment by ajafff** [comment]
That indeed looks like a bug. Do you have strictNullChecks enabled in your tsconfig.json?

**Comment by Martin-Wegner** [comment]
No `strictNullChecks` is not enabled.

**Comment by ajafff** [comment]
This rule does not work as expected without strictNullChecks, because the current implementation assumes that it's enabled.

I doubt it would be useful to add special handling for `strictNullChecks = false`, because the purpose of this rule is to check if you are comparing to `null` or `undefined` on non-nullable values. Everything else is already checked by the compiler.
The only check that could be performed without strictNullChecks would be somethink like this:
```ts
declare var num: number;
typeof num === "function"; // this will never be true
typeof num === "symbol"; // this will never be true
typeof num === "string"; // this will never be true
...
typeof num === "object"; // could be true when num === null
typeof num === "undefined"; // could be true when num === undefined
```

**Comment by ajafff** [comment]
Of course we could do a better job documenting that behavior. Maybe even warn at runtime if this rule is used without strictNullChecks (and not apply this rule?).
